packages/dds/tree/src/core/tree/anchorSet.ts

```diff
@@ -542,7 +511,7 @@ export class AnchorSet implements ISubscribable<AnchorSetRootEvents> {
 	public applyDelta(delta: Delta.Root): void {
```

Comment:
It should have been possible to fix this bug without changing moveChildren since as it was it was flexible enough to express both attach and detach (move them to a detached sequence, just like how object forest implements its delta application).

With the approach you have here, It seems like it complicates reference counting / ownership a bit, since in-between attach and detach the anchorsNodes are in a special state where they don't have an owner, and you have to define what this means for their reference counts.

Also that approach avoids having this code (applyDelta) have to think about index adjustments since the maintainace of invariance (including indexes, ref count and parent pointers) is all encapsulated into moveChildren.

It has the further benefit of making this code work more like the code in the forests, which may make them more maintainable if/when we change how deltas work and or want to unify some of the code.

Its unclear to me if the refactoring to not use moveChildren as is was done because this new pattern is believed to be better, or out of lack of clarity on how to use it as it was: was this change intended as an improvement? (I'd like to know the intend here before reviewing the details of it).

Reply:
I splitted the move method for this reasons:
1) When we are trying to apply moveOut and moveIn we can't do the detach and attach anchors nodes in the same call(That's one of the reasons of the bug).
2) It helps with readibility and apply the single responsability design principle.
	-When there is an insertion of nodes in the original tree, there is not move of nodes in the anchorSet, just affected nodes need to increase parentIndex.
	-When there is a delete of nodes in the original tree, there is not move of nodes in the anchorSet, just those nodes are detached and marked as deleted.

Reply:
I splitted the move method for this reasons:
1) When we are trying to apply moveOut and moveIn we can't do the detach and attach anchors nodes in the same call(That's one of the reasons of the bug).
2) It helps with readibility and apply the single responsability design principle.
	-When there is an insertion of nodes in the original tree, there is not move of nodes in the anchorSet, just affected nodes need to increase parentIndex.
	-When there is a delete of nodes in the original tree, there is not move of nodes in the anchorSet, just those nodes are detached and marked as deleted.

Reply:
1. You would call move twice. Move out is just a move to a new detached sequence, and move in is just a move from there to the destination.
2. With your changes, applyDelta is no longer just responsible for moving things around, it now has to understand the implementation details of anchorset. More specifically you are leaving the anchoset in an inconstant state between a move out and a move in (the content which is moved out doesn't have valid indexes, or a valid parent, and you have references to those anchors which arn't using the ref counting), so everything running in the window has to know the implementation details of anchorset and make sure they don't mess anything up. For example, if one of the events applyDelta invokes removes its event handler, that could result in the anchor deleting itself. If that was an anchor that was moved out but not moved in, it deleting itself will not work correctly. Fully putting the anchorset into a con sistant state (where edits like adding and removing anchors will work correctly) when a moveout occurs (ex: by using the old move function) solves this. I think this abstraction (that apply delta and all event handlers it invokes don't have to worry about ref counting and other internals of anchorset) is more important than logically separating the two halves of move.

That said, I'm not convinced even the approach using two moves actually fully handles events which free anchors correctly (since applyDelta doesn't keep a ref count to the path its at while walking), but its much closer to robust in that respect.

It might make sense to move applyDelta to be a free function and implement it in terms of the AnchorSet's (the class)'s public API to help make it more explicit what the invariants are during applyDelta and its event callbacks.

Reply:
If you do end up going with the moveOut to detached sequence approach, you can keep the splitting up of move into separate parts that you did (and provide a move method that wraps them), but they will have to carefully document the state they leave things in.